### ISSUE TITLE
Enforce balanced parentheses (V2)

### ISSUE THREAD

**Issue by paustint** [paustint/soql-parser-js#80]
- [ ] Feature
- [ ] Bug Request

## Description

1. Enforce balanced parentheses or throw an error
1. Ensure that any variation of nested parentheses is supported
1. Add unit test-cases including more parentheses examples
1. Add unit tests for invalid queries that focus on where clauses

Example - this does not break:
`SELECT Id, Owner.Name FROM Task WHERE ((Owner.FirstName LIKE 'B%')`

**Comment by paustint** [comment]
![image](https://user-images.githubusercontent.com/5461649/66070872-5a135f80-e50f-11e9-8160-c8e6cbeefa34.png)

do you have any design ideas on the best way to ensure that parentheses are balanced?
trivial example: (foo = bar && (bar = baz || bar = zoo))
Let's say the end paren is missing, I would want to throw an error during parsing (if possible).
Right now I am passing arguments to rules so that I keep track of my opening parens and use a gate to make sure I do not over-consume closing parens - but that does not solve the case where there are too few closing parens.
And I am not sure if my strategy is the best, or if there are other ways that people handle these types of things that I am not aware of.

Matthew Dean @matthew-dean 12:06
@paustint You should be able to do that with a rule that's like:

this.RULE('parenBlock', () => {
  this.CONSUME(LParen)
  this.SUBRULE(...)
  this.CONSUME(RParen)
})
And then in your subrule, one of the OR things it can consume is another parenBlock

so like:
this.OR([
  { ALT: () => this.SUBRULE(this.expression) },
  { ALT: () => this.SUBRULE(this.parenBlock) }
])
i.e. just make the parenBlock recursive, and it will match all start / end parentheses as a result

Matthew Dean @matthew-dean 12:14
I wouldn't "keep track" of parens; the best parsing strategy is usually just to rely on recursion to eventually close (match) the end parens from each time you entered the block
just to illustrate this more clearly, a simplified example would be:
this.RULE('parenBlock', () => {
  this.CONSUME(LParen)
  this.OR([
    { ALT: () => this.SUBRULE(this.expression) },
    { ALT: () => this.SUBRULE(this.parenBlock) }
  ])
  this.CONSUME(RParen)
})

Austin Turner @paustint 12:16
What about this:
(((((foo = bar)))))

Matthew Dean @matthew-dean 12:16
Same thing, look at the example
It will keep entering the parenBlock as the matching path of the OR

Austin Turner @paustint 12:17
Oh, I see

Matthew Dean @matthew-dean 12:17
but it won't move on to the next statement this.CONSUME(RParen) until it's exited

Austin Turner @paustint 12:18
Cool - this is very helpful, thank you so much!

Matthew Dean @matthew-dean 12:18
so it basically recurses deeper and deeper, and then when the expression is finally consumed, it then goes on to the next CONSUME(RParen), then that rule is complete, then goes up to the next parenBlock and runs CONSUME(RParen) on that one, etc

Austin Turner @paustint 12:19


Matthew Dean @matthew-dean 12:19
np, the Chevrotain examples are very good for this, at first I was trying to do a lot of "tracking" but often the right kind of nesting / recursion can reduce this logic
Basically, try to give Chevrotain a limited set of choices, and reduce GATEs if you can

Austin Turner @paustint 12:22
Thank you for the advice.
I have my LParen and RParen inside a many function, so there is no control over the matching of expressions - I will be really happy to refactor that bad design out of my code ;)

Matthew Dean @matthew-dean 12:22
oh yeah no, that sounds bad lol
